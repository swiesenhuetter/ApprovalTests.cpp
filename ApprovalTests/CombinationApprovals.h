#pragma once

#include <type_traits>
#include "reporters/DefaultReporter.h"
#include "ApprovalTests/core/Reporter.h"
#include "ApprovalTests/utilities/CartesianProduct.h"
#include "Approvals.h"

namespace ApprovalTests
{
    namespace CombinationApprovals
    {
        namespace Detail
        {

            // Write out second or subsequent input value, with preceding comma and space
            struct print_input
            {
                std::ostream& out;
                template <class T> void operator()(const T& input)
                {
                    out << ", " << input;
                }
            };

            // Write out one row of output
            template <class Converter> struct serialize
            {
                std::ostream& out;
                Converter converter;
                template <class T, class... Ts>
                void operator()(T&& input1_, Ts&&... inputs)
                {
                    // First value is printed without trailing comma
                    out << "(" << input1_;
                    // Remaining values are printed with prefix of a comma
                    CartesianProduct::Detail::for_each(std::forward_as_tuple(inputs...),
                                                       print_input{out});
                    out << ") => " << converter(input1_, inputs...) << '\n';
                }
            };
        } // namespace Detail

        template <class Converter, class Container, class... Containers>
        void verifyAllCombinations(const Reporter& reporter,
                                   Converter&& converter,
                                   const Container& input0,
                                   const Containers&... inputs)
        {
            std::stringstream s;
            CartesianProduct::cartesian_product(
                Detail::serialize<Converter>{s, std::forward<Converter>(converter)},
                input0,
                inputs...);
            Approvals::verify(s.str(), reporter);
        }

        template <class Converter, class... Containers>
        ApprovalTests::Detail::EnableIfNotDerivedFromReporter<Converter>
        verifyAllCombinations(Converter&& converter, const Containers&... inputs)
        {
            verifyAllCombinations(
                DefaultReporter(), std::forward<Converter>(converter), inputs...);
        }

    } // namespace CombinationApprovals
} // namespace ApprovalTests
