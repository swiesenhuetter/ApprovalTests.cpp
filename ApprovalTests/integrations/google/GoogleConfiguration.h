#pragma once

#include "GoogleCustomizationsFactory.h"

namespace ApprovalTests
{
    class GoogleConfiguration
    {
    public:
        // This result is not used, it is only there to allow the method to execute, when this is used outside a function.
        APPROVAL_TESTS_NO_DISCARD static bool
        addTestCaseNameRedundancyCheck(GoogleCustomizationsFactory::Comparator comparator)
        {
            return GoogleCustomizationsFactory::addTestCaseNameRedundancyCheck(
                comparator);
        }

        // This result is not used, it is only there to allow the method to execute, when this is used outside a function.
        APPROVAL_TESTS_NO_DISCARD static bool
        addIgnorableTestCaseNameSuffix(std::string suffix)
        {
            return addTestCaseNameRedundancyCheck(
                createIgnorableTestCaseNameSuffixCheck(suffix));
        }

        static GoogleCustomizationsFactory::Comparator
        createIgnorableTestCaseNameSuffixCheck(const std::string& suffix)
        {
            return [suffix](std::string testFileNameWithExtension,
                            std::string testCaseName) {
                if (testCaseName.length() <= suffix.length() ||
                    !StringUtils::endsWith(testCaseName, suffix))
                {
                    return false;
                }

                auto withoutSuffix =
                    testCaseName.substr(0, testCaseName.length() - suffix.length());
                auto withFileExtension = withoutSuffix + ".";
                return StringUtils::contains(testFileNameWithExtension,
                                             withFileExtension);
            };
        }
    };
}
