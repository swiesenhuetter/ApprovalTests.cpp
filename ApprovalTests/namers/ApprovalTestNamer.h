#pragma once

#include "ApprovalTests/core/ApprovalNamer.h"
#include <sstream>
#include <vector>
#include <stdexcept>
#include "ApprovalTests/utilities/SystemUtils.h"
#include "ApprovalTests/namers/HelpMessages.h"

namespace ApprovalTests
{
    class TestName
    {
    public:
        const std::string& getFileName() const
        {
            checkBuildConfiguration(fileName);
            return fileName;
        }

        void setFileName(const std::string& file)
        {
            fileName = SystemUtils::checkFilenameCase(file);
        }

    private:
        static void checkBuildConfiguration(const std::string& fileName)
        {
            if (!FileUtils::fileExists(fileName))
            {
                throw std::runtime_error(getMisconfiguredBuildHelp(fileName));
            }
        }

    public:
        static std::string getMisconfiguredBuildHelp(const std::string& fileName)
        {
            return "\n\n" + HelpMessages::getMisconfiguredBuildHelp(fileName) + "\n\n";
        }

        std::vector<std::string> sections;

    private:
        std::string fileName;
    };

    class TestConfiguration
    {
    public:
        std::string subdirectory;
    };

    class ApprovalTestNamer : public ApprovalNamer
    {
    private:
    public:
        ApprovalTestNamer() = default;

        std::string getTestName() const
        {
            std::stringstream ext;
            auto test = getCurrentTest();
            for (size_t i = 0; i < test.sections.size(); i++)
            {
                if (0 < i)
                {
                    ext << ".";
                }
                ext << test.sections[i];
            }

            return convertToFileName(ext.str());
        }

        static bool isForbidden(char c)
        {
            static std::string forbiddenChars("\\/:?\"<>|' ");
            return std::string::npos != forbiddenChars.find(c);
        }

        static std::string convertToFileName(const std::string& fileName)
        {
            std::stringstream result;
            for (auto ch : fileName)
            {
                if (!isForbidden(ch))
                {
                    result << ch;
                }
                else
                {
                    result << "_";
                }
            }
            return result.str();
        }

        static TestName& getCurrentTest()
        {
            try
            {
                return currentTest();
            }
            catch (const std::runtime_error&)
            {
                std::string helpMessage = getMisconfiguredMainHelp();
                throw std::runtime_error(helpMessage);
            }
        }

        static std::string getMisconfiguredMainHelp()
        {
            return "\n\n" + HelpMessages::getMisconfiguredMainHelp() + "\n\n";
        }

        // Deprecated - please use getSourceFileName
        std::string getFileName() const
        {
            return getSourceFileName();
        }

        std::string getSourceFileName() const
        {
            auto file = getCurrentTest().getFileName();
            auto start = file.rfind(SystemUtils::getDirectorySeparator()) + 1;
            auto end = file.rfind('.');
            auto fileName = file.substr(start, end - start);
            return convertToFileName(fileName);
        }

        std::string getDirectory() const
        {
            auto file = getCurrentTest().getFileName();
            auto end = file.rfind(SystemUtils::getDirectorySeparator()) + 1;
            auto directory = file.substr(0, end);
            if (!testConfiguration().subdirectory.empty())
            {
                directory += testConfiguration().subdirectory +
                             SystemUtils::getDirectorySeparator();
                SystemUtils::ensureDirectoryExists(directory);
            }
            return directory;
        }

        static TestName& currentTest(TestName* value = nullptr)
        {
            static TestName* staticValue;
            if (value != nullptr)
            {
                staticValue = value;
            }
            if (staticValue == nullptr)
            {
                throw std::runtime_error(
                    "The variable in currentTest() is not initialised");
            }
            return *staticValue;
        }

        static TestConfiguration& testConfiguration()
        {
            static TestConfiguration configuration;
            return configuration;
        }

        virtual std::string getApprovedFile(std::string extensionWithDot) const override
        {

            return getFullFileName(".approved", extensionWithDot);
        }

        virtual std::string getReceivedFile(std::string extensionWithDot) const override
        {

            return getFullFileName(".received", extensionWithDot);
        }

        std::string getOutputFileBaseName() const
        {
            return getSourceFileName() + "." + getTestName();
        }

        std::string getFullFileName(const std::string& approved,
                                    const std::string& extensionWithDot) const
        {
            std::stringstream ext;
            ext << getDirectory() << getOutputFileBaseName() << approved
                << extensionWithDot;
            return ext.str();
        }
    };
}
