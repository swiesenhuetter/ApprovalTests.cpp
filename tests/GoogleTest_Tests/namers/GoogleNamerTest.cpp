#include "gtest/gtest.h"
#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/CombinationApprovals.h"
#include "ApprovalTests/integrations/google/GoogleConfiguration.h"

#include <ostream>
#include <string>

using namespace ApprovalTests;

TEST(GoogleNamerTest, ItDropsFirstNameWhenItEqualsTheFilename)
{
    ApprovalTestNamer namer;
    EXPECT_EQ(namer.getTestName(), "ItDropsFirstNameWhenItEqualsTheFilename");
}

TEST(TestCaseNameDifferentNameThanFile, TestName)
{
    ApprovalTestNamer namer;
    EXPECT_EQ(namer.getTestName(), "TestCaseNameDifferentNameThanFile.TestName");
}

std::string createSuffix(const std::string& suffix,
                         const std::string& fileName,
                         const std::string& testCaseName)
{
    auto converter = GoogleConfiguration::createIgnorableTestCaseNameSuffixCheck(suffix);
    return converter(fileName, testCaseName) ? "redundant" : "";
}

TEST(GoogleNamerTest, TestSuffixMatcher)
{
    std::string suffix = "Test";
    std::string fileName = "/a/b/c/testGoogleNamer.cpp";
    std::vector<std::string> testCaseNames = {"GoogleNamerTest",
                                              "GoogleNamer",
                                              "GoogleTest",
                                              "NamerTest",
                                              "NamerTestTest",
                                              "NamerTestTests",
                                              "TestTest",
                                              "Test"};
    Approvals::verifyAll<std::vector<std::string>>(
        "suffix: " + suffix + "\nfilename: " + fileName + "\ntest case names:",
        testCaseNames,
        [&](const std::string& test, std::ostream& os) {
            os << test << ": " << createSuffix(suffix, fileName, test);
        });
}

TEST(GoogleNamerTest, TestSuffixMatcherBug)
{
    std::string testCaseName = "ApprovalTestsTest";
    std::string suffix = "Test";
    std::string fileName = "/a/b/c/testApprovalTests.cpp";
    auto converter = GoogleConfiguration::createIgnorableTestCaseNameSuffixCheck(suffix);
    EXPECT_EQ(converter(fileName, testCaseName), true);
}
