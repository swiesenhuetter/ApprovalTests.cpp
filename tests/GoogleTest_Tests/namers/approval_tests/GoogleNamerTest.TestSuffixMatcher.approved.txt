suffix: Test
filename: /a/b/c/testGoogleNamer.cpp
test case names:


GoogleNamerTest: redundant
GoogleNamer: 
GoogleTest: 
NamerTest: redundant
NamerTestTest: 
NamerTestTests: 
TestTest: 
Test: 

